cf_xarray/accessor.py

```diff
@@ -170,6 +180,69 @@ def _get_groupby_time_accessor(var: Union[DataArray, Dataset], key: str) -> List
         return []
 
 
+def _get_custom_criteria(
+    obj: Union[DataArray, Dataset], key: str, criteria=None
+) -> List[str]:
+    """
+    Translate from axis, coord, or custom name to variable name optionally
+    using ``custom_criteria``
+
+    Parameters
+    ----------
+    obj : DataArray, Dataset
+        DataArray belonging to the coordinate to be checked
+    key : str, ["X", "Y", "Z", "T", "longitude", "latitude", "vertical", "time"], or a key from user-defined custom_criteria
+        key to check for.
+    criteria : dict, optional
+        Criteria to use to map from variable to attributes describing the
+        variable. An example is coordinate_criteria which maps coordinates to
+        their attributes and attribute values. If user has defined
+        custom_criteria, this will be used by default.
+
+    Returns
+    -------
+    List[str], Variable name(s) in parent xarray object that matches axis, coordinate, or custom `key`
+
+    """
+
+    if isinstance(obj, DataArray):
+        obj = obj._to_temp_dataset()
+
+    if criteria is None:
+        if not OPTIONS["custom_criteria"]:
+            return []
+        criteria = OPTIONS["custom_criteria"]
+
+    if criteria is not None:
+        criteria = always_iterable(criteria, allowed=(tuple, list, set))
+
+    criteria = ChainMap(*criteria)
+
+    results: Set = set()
+    if key in criteria:
+        for criterion, expected in criteria[key].items():
+            for var in obj.variables:
+                # Treat expected as regex
+                expected_vals = [*expected.values()][0]
+                # use regex to match
+                if "regex" in expected:
+                    if re.match(
+                        "|".join(expected_vals), obj[var].attrs.get(criterion, "")
+                    ):
+                        results.update((var,))
+                    # also check name specifically since not in attributes
+                    elif criterion == "name" and re.match("|".join(expected_vals), var):
+                        results.update((var,))
+                else:  # exact matches only
+                    if obj[var].attrs.get(criterion, "") in expected_vals:
+                        # if re.match("|".join(expected), obj[var].attrs.get(criterion, "")):
+                        results.update((var,))
+                    # also check name specifically since not in attributes
+                    elif criterion == "name" and var in expected_vals:
+                        results.update((var,))
```

Comment:
Lets raise an error here if kind is not "regex" or "exact" (and also test for that error)

Reply:
This is in https://github.com/xarray-contrib/cf-xarray/pull/234/commits/e15a2a0adb6367aae0ed122bf79e322b22a6d51c